### ISSUE TITLE
NSFetchedResultsController breaking down

### ISSUE THREAD

**Issue by flypaper0** [DenHeadless/DTTableViewManager#52]
Hello, Den! Thanks for your library. 

When I'm trying to add and update CoreData Item in single transaction, NSFetchedResultsController is break down. App not crashing, but data in tableview is incorrect, and no ways to reload, or update.

[Demo project.zip](https://www.dropbox.com/s/6ktsv8w4rcbd1rw/tableviewbug.zip?dl=0)

**Comment by DenHeadless** [comment]
Hey! 

Thanks for demo project, it was very helpful to track issue down.

This is actually a documented behaviour of `NSFetchedResultsController`(https://developer.apple.com/reference/coredata/nsfetchedresultscontroller):

```
Changes are not reflected until after the controller’s managed object context has received a processPendingChanges() message. Therefore, if you change the value of a managed object’s attribute so that its location in a fetched results controller’s results set would change, its index as reported by the controller would typically not change until the end of the current event cycle (when processPendingChanges() is invoked).
```

What this means is when you do insertions/removals/moves of your data models in CoreData, and you want to simultaneously update data models, you should call processPendingChanges() on your `NSManagedObjectContext` prior to updating them, or you will receive wrong IndexPaths reported by `NSFetchedResultsController`.

For example, to fix example project you provided, just modify doit method to look like this:

```swift
func doit() {
    insertItems(("B", "1"), ("B", "2"), ("C", "1"))
    
    moc.processPendingChanges()
    
    let request = NSFetchRequest<Item>(entityName: "Item")
    request.predicate = NSPredicate(format: "section = %@ AND row = %@", "D", "3")
    if let item = try! moc.fetch(request).first {
        item.row = "2!!!"
    }
    
    try! moc.save()
}
```

Hope this helps in solving your issue.

**Comment by flypaper0** [comment]
@DenHeadless thanks for quick feedback. In our project impossible divide insertItems transaction and update, like in demo project.

So the only one way to solve this problem is (138 line CoreDataStorage):

```
case .update:
            if let indexPath = indexPath {
                if let new = newIndexPath {
                    currentUpdate?.objectChanges.append((.delete,[indexPath]))
                    currentUpdate?.objectChanges.append((.insert,[new]))
                } else {
                    currentUpdate?.objectChanges.append((.update,[indexPath]))
                }
                currentUpdate?.updatedObjects[indexPath] = anObject
            }
        }
```

What do you think?

**Comment by DenHeadless** [comment]
How do you think this can be fixed in CoreDataStorage?

**Comment by flypaper0** [comment]
@DenHeadless opened pull request in DTModelStorage (#17) 
[update + insert in one transaction iOS 10 fix #17](https://github.com/DenHeadless/DTModelStorage/pull/17)
